### PR TITLE
Add `device_matrix_data` and device-side `matrix::read`

### LINKED ISSUE #729: Repeated allocations while reading `Fbcsr` matrix
Currently, the `Fbcsr` `read` function allocates one `std::vector` for every small block. This may impact read performance for the smallest block sizes.

Since `read` is typically called very few times per program execution, this is low-priority.

Related pull-request: https://github.com/ginkgo-project/ginkgo/pull/671

### PR DESCRIPTION
**Pull request by upsj**
This PR adds a new `device_matrix_data` that works equivalently to `matrix_data` except that it uses a device-side `Array`. I provide a default implementation using the old host-side initialization.

There are still some things to do, but I wanted to ask for comments already.

- [x] merge #833 to implement Ell `compute_max_nnz`
- [x] add `Ell::read`
- [x] add `Hybrid::read`
- [x] add `Sellp::read`
- [x] ~~modify Ell/Sellp sentinel values~~ #905
- [x] ~~modify conversions for numerical zeros~~ This will be taken care of with #905 
- [x] ~~move `read` tests from `core` to `reference`~~ (this will no longer be necessary with #889)
- [x] ~~add tests checking zero behavior~~ This will be taken care of with #904
- [x] add tests for `remove_zeros` and device_matrix_data functionality

Fixes #729

### CODE REVIEW

**Review by yhmtsai — COMMENTED** [review]
First part of my review.
It does not contain hybrid/sellp/fb_csr implementation.
Some comments containing Reminder is a note for me because it is hard to go different file on GitHub review.

File: common/unified/matrix/ell_kernels.cpp [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+        [] GKO_KERNEL(auto row, auto nonzeros, auto row_ptrs, auto stride,
+                      auto num_cols, auto cols, auto values) {
+            const auto begin = row_ptrs[row];
+            const auto end = row_ptrs[row + 1];
+            const auto nnz = end - begin;
+            auto out_idx = row;
+            for (auto i = begin; i < end; i++) {
+                const auto nonzero = nonzeros[i];
+                cols[out_idx] = nonzero.column;
+                values[out_idx] = unpack_member(nonzero.value);
```
[the remark this answers is not in the document]
Comment (upsj) [reply common/unified/matrix/ell_kernels.cpp:90]: this workaround is only necessary for structs containing std/thrust::complex members, since its alignment restrictions cause the struct to have a different size between GPU and CPU. So this is specific to members, not values.

File: common/unified/matrix/ell_kernels.cpp [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+                out_idx += stride;
+            }
+            for (auto i = nnz; i < num_cols; i++) {
+                cols[out_idx] = 0;
+                values[out_idx] = zero(values[out_idx]);
```
[the remark this answers is not in the document]
Comment (upsj) [reply common/unified/matrix/ell_kernels.cpp:95]: The function takes a reference, so the values doesn't actually need to be read. Another solution could be `zero<device_type<ValueType>>()`, I guess?

File: common/unified/matrix/ell_kernels.cpp [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+            const auto nnz = end - begin;
+            auto out_idx = row;
+            for (auto i = begin; i < end; i++) {
+                const auto nonzero = nonzeros[i];
+                cols[out_idx] = nonzero.column;
+                values[out_idx] = unpack_member(nonzero.value);
+                out_idx += stride;
+            }
+            for (auto i = nnz; i < num_cols; i++) {
+                cols[out_idx] = 0;
```
[the remark this answers is not in the document]
Comment (upsj) [reply common/unified/matrix/ell_kernels.cpp:94]: That will change in another PR, I wanted to keep the fundamental changes out of this PR

File: core/base/device_matrix_data.cpp [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+    std::shared_ptr<const Executor> exec, dim<2> size, size_type nnz)
+    : size{size}, nonzeros{exec, nnz}
+{}
+
+
+template <typename ValueType, typename IndexType>
+device_matrix_data<ValueType, IndexType>::device_matrix_data(
+    dim<2> size, Array<nonzero_type> data)
+    : size{size}, nonzeros{std::move(data)}
+{}
```
[the remark this answers is not in the document]
Comment (upsj) [reply core/base/device_matrix_data.cpp:66]: I am trying to reduce the amount of function implementations we have in our headers, since they make it harder to find important things, and make the compilation more costly.

File: core/base/device_matrix_data.cpp [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+
+
+template <typename ValueType, typename IndexType>
+device_matrix_data<ValueType, IndexType>
+device_matrix_data<ValueType, IndexType>::create_from_host(
+    std::shared_ptr<const Executor> exec, host_type& data)
+{
+    auto host_view = Array<nonzero_type>::view(
+        exec->get_master(), data.nonzeros.size(), data.nonzeros.data());
+    auto device_view = Array<nonzero_type>{exec, std::move(host_view)};
```
[the remark this answers is not in the document]
Comment (upsj) [reply core/base/device_matrix_data.cpp:90]: I'm trying to keep the overhead of create_from_host minimal when exec is a CPU executor. But you make a good point - it should probably have `view` in its name, or remove the `std::move` call in the constructor of the resulting object to copy the data.

File: core/matrix/coo.cpp [hunk 6]
```diff
@@ -193,36 +194,29 @@ void Coo<ValueType, IndexType>::move_to(Dense<ValueType>* result)
...
+    this->row_idxs_.resize_and_reset(nnz);
+    this->col_idxs_.resize_and_reset(nnz);
+    this->values_.resize_and_reset(nnz);
+    auto nonzeros = make_temporary_clone(exec, &data.nonzeros);
+    exec->run(coo::make_from_matrix_data(*nonzeros, this));
 }
 
 
 template <typename ValueType, typename IndexType>
 void Coo<ValueType, IndexType>::write(mat_data& data) const
```
[the remark this answers is not in the document]
Comment (upsj) [reply core/matrix/coo.cpp:217]: I don't think that is worth it, `write` performance seems much less necessary to me than `read` performance.

File: core/matrix/dense.cpp [hunk 7]
```diff
@@ -670,43 +671,49 @@ void Dense<ValueType>::move_to(SparsityCsr<ValueType, int64>* result)
...
-                data.nonzeros[ind].column == col) {
-                tmp->at(row, col) = data.nonzeros[ind].value;
-                ++ind;
-            } else {
-                tmp->at(row, col) = zero<typename MatrixType::value_type>();
-            }
-        }
+    if (this->get_size() != data.size) {
+        this->set_size(data.size);
+        this->stride_ = data.size[1];
```
[the remark this answers is not in the document]
Comment (upsj) [reply core/matrix/dense.cpp:679]: My intention here was to avoid reallocations if possible in all cases. That makes interaction with array views much simpler

File: core/matrix/sparsity_csr.cpp [hunk 8]
```diff
@@ -95,33 +98,27 @@ void SparsityCsr<ValueType, IndexType>::apply_impl(const LinOp* alpha,
...
 
+template <typename ValueType, typename IndexType>
+void SparsityCsr<ValueType, IndexType>::read(const device_mat_data& data)
+{
+    const auto nnz = data.nonzeros.get_num_elems();
+    auto exec = this->get_executor();
+    this->set_size(data.size);
+    this->row_ptrs_.resize_and_reset(data.size[0] + 1);
+    this->col_idxs_.resize_and_reset(nnz);
+    this->value_.fill(one<ValueType>());
```
[the remark this answers is not in the document]
Comment (upsj) [reply core/matrix/sparsity_csr.cpp:109]: That's the current behavior for read, IIRC

File: dpcpp/CMakeLists.txt [hunk 9]
```diff
@@ -54,7 +55,7 @@ target_sources(ginkgo_dpcpp
     )
 
 ginkgo_compile_features(ginkgo_dpcpp)
-target_compile_definitions(ginkgo_dpcpp PRIVATE GKO_COMPILING_DPCPP)
+target_compile_definitions(ginkgo_dpcpp PRIVATE GKO_COMPILING_DPCPP _ONEDPL_COMPILE_KERNEL=0)
```
[the remark this answers is not in the document]
Comment (upsj) [reply dpcpp/CMakeLists.txt:58]: I think we should wait for Alina's feedback when the oneDPL folks got back to her.

File: include/ginkgo/core/base/range.hpp [hunk 10]
```diff
@@ -106,11 +106,7 @@ struct span {
      *
      * @return `this->end - this->begin`
      */
-    GKO_ATTRIBUTES constexpr size_type length() const
-    {
-        GKO_ASSERT(is_valid());
```
[the remark this answers is not in the document]
Comment (upsj) [reply include/ginkgo/core/base/range.hpp:111]: adding the assertion here breaks ROCm compilation due to some weird include order issue after adding the device_matrix_data include to types.hpp

File: include/ginkgo/core/matrix/fbcsr.hpp [hunk 11]
```diff
@@ -300,25 +303,21 @@ class Fbcsr : public EnableLinOp<Fbcsr<ValueType, IndexType>>,
...
-     */
-    void set_block_size(const int block_size) noexcept { bs_ = block_size; }
-
     /**
      * @return The number of block-rows in the matrix
      */
     index_type get_num_block_rows() const noexcept
     {
-        return row_ptrs_.get_num_elems() - 1;
+        return this->get_size()[0] / bs_;
```
[the remark this answers is not in the document]
Comment (upsj) [reply include/ginkgo/core/matrix/fbcsr.hpp:311]: size is always divisible by block size

File: include/ginkgo/core/matrix/fbcsr.hpp [hunk 12]
```diff
@@ -300,25 +303,21 @@ class Fbcsr : public EnableLinOp<Fbcsr<ValueType, IndexType>>,
      */
     int get_block_size() const noexcept { return bs_; }
 
-    /**
-     * Set the fixed block size for this matrix
-     *
-     * @param block_size The block size
-     */
-    void set_block_size(const int block_size) noexcept { bs_ = block_size; }
```
[the remark this answers is not in the document]
Comment (upsj) [reply include/ginkgo/core/matrix/fbcsr.hpp:308]: Adding this function was a mistake, as it allows users to completely break the Fbcsr type. Any use on a non-empty matrix causes any subsequent operations to either produce wrong results or (more likely) segfault.

File: include/ginkgo/core/matrix/sellp.hpp [hunk 13]
```diff
@@ -275,8 +281,8 @@ class Sellp : public EnableLinOp<Sellp<ValueType, IndexType>>,
     /**
      * @copydoc Sellp::col_at(size_type, size_type, size_type)
      */
-    index_type col_at(size_type row, size_type slice_set, size_type idx) const
-        noexcept
+    index_type col_at(size_type row, size_type slice_set,
+                      size_type idx) const noexcept
```
[the remark this answers is not in the document]
Comment (upsj) [reply include/ginkgo/core/matrix/sellp.hpp:285]: yep, I'll try to revert that before merging.

File: include/ginkgo/core/matrix/sellp.hpp [hunk 14]
```diff
@@ -328,8 +334,7 @@ class Sellp : public EnableLinOp<Sellp<ValueType, IndexType>>,
           slice_lengths_(exec, ceildiv(size[0], slice_size)),
           slice_sets_(exec, ceildiv(size[0], slice_size) + 1),
           slice_size_(slice_size),
-          stride_factor_(stride_factor),
-          total_cols_(total_cols)
```
[the remark this answers is not in the document]
Comment (upsj) [reply include/ginkgo/core/matrix/sellp.hpp:332]: total_cols can be computed from the size of `values` and `col_idxs`, so we don't need to store it.

File: core/matrix/diagonal.cpp [hunk 15]
```diff
@@ -234,17 +235,45 @@ inline void read_impl(MatrixType* mtx, const MatrixData& data)
...
+    auto exec = this->get_executor();
+    exec->run(diagonal::make_fill_in_matrix_data(
+        *make_temporary_clone(exec, &data.nonzeros), this));
+}
+
+
+template <typename ValueType>
+void Diagonal<ValueType>::read(const device_mat_data32& data)
+{
+    GKO_ASSERT_IS_SQUARE_MATRIX(data.size);
```
[the remark this answers is not in the document]
Comment (upsj) [reply core/matrix/diagonal.cpp:254]: As we don't have dense::convert_to(diagonal), so I would keep these operations conceptually separate.

File: common/unified/matrix/ell_kernels.cpp [hunk 16]
```diff
@@ -0,0 +1,111 @@
...
+            const auto end = row_ptrs[row + 1];
+            const auto nnz = end - begin;
+            auto out_idx = row;
+            for (auto i = begin; i < end; i++) {
+                const auto nonzero = nonzeros[i];
+                cols[out_idx] = nonzero.column;
+                values[out_idx] = unpack_member(nonzero.value);
+                out_idx += stride;
+            }
+            for (auto i = nnz; i < num_cols; i++) {
```
[the remark this answers is not in the document]
Comment (upsj) [reply common/unified/matrix/ell_kernels.cpp:93]: right, I did the same thing for Sellp, will do that.

File: common/unified/matrix/ell_kernels.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (yhmtsai) [reply common/unified/matrix/ell_kernels.cpp:95]: > The function takes a reference, so the values doesn't actually need to be read. 

Yes, I think so, too. the function only uses the type of value to decide which one to run.
> Another solution could be `zero<device_type<ValueType>>()`, I guess?

Yes

File: common/unified/matrix/ell_kernels.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (yhmtsai) [reply common/unified/matrix/ell_kernels.cpp:94]: I see.

File: core/base/device_matrix_data.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (yhmtsai) [reply core/base/device_matrix_data.cpp:90]: after reviewing more files, I realize the executor can also be host, so it is view actually when executor is host

File: common/cuda_hip/matrix/fbcsr_kernels.hpp.inc [hunk 17]
```diff
@@ -73,3 +73,86 @@ __global__ __launch_bounds__(default_block_size) void transpose_blocks(
...
+    Array<ValueType>& value_array)
+{
+    using entry = matrix_data_entry<ValueType, IndexType>;
+    using device_entry = device_type<entry>;
+    const auto nnz = data.get_num_elems();
+    const auto bs = block_size;
+    auto row_ptrs = row_ptr_array.get_data();
+    auto num_rows = row_ptr_array.get_num_elems() - 1;
+    if (nnz == 0) {
+        components::fill_array(exec, row_ptrs, num_rows, IndexType{});
```
Comment (thoasm) [inline common/cuda_hip/matrix/fbcsr_kernels.hpp.inc:92]: Shouldn't this be `num_rows + 1`?
```suggestion
        components::fill_array(exec, row_ptrs, num_rows + 1, IndexType{});
```

**Review by thoasm — APPROVED** [review]
Mostly nits, so LGTM!

File: cuda/base/types.hpp [hunk 18]
```diff
@@ -331,6 +397,27 @@ inline culibs_type<T> as_culibs_type(T val)
...
 
 
+/**
+ * Casts fake_complex<T> to thrust::complex<T> and leaves any other types
+ * unchanged.
+ *
+ * This is necessary to work around an issue with Thrust shipped in CUDA 9.2,
+ * and the fact that thrust::complex has stronger alignment restrictions than
+ * std::complex, i.e. structs containing them among other smaller members have
+ * different sizes on device and host.
```
[the remark this answers is not in the document]
Comment (upsj) [reply cuda/base/types.hpp:407]: yes, though with using as_cuda_type, that means effectively the same thing.

File: common/unified/matrix/sparsity_csr_kernels.cpp [hunk 19]
```diff
@@ -0,0 +1,71 @@
...
+/**
+ * @brief The SparsityCsr matrix format namespace.
+ *
+ * @ingroup sparsity_csr
+ */
+namespace sparsity_csr {
+
+
+template <typename ValueType, typename IndexType>
+void from_matrix_data(
```
[the remark this answers is not in the document]
Comment (upsj) [reply common/unified/matrix/sparsity_csr_kernels.cpp:51]: how about using fill_in_matrix_data consistently everywhere? The issue is really that each matrix format is slightly different, so I tried to use a consistent name for the part that actually finishes the matrix data read, except for Diagonal and Dense, where it just fills in the non-zeros into an already zeroed array.

File: common/cuda_hip/matrix/fbcsr_kernels.hpp.inc (hunk 17 above)
Comment (upsj) [reply common/cuda_hip/matrix/fbcsr_kernels.hpp.inc:92]: Yep, due to all initializations going through Reference, we aren't actually testing the device-side read functions right now. Working on fixing that.

File: cuda/base/types.hpp [hunk 20]
```diff
@@ -57,6 +60,53 @@ namespace cuda {
 namespace detail {
 
 
+/**
+ * @internal
+ *
+ * replacement for thrust::complex without alignment restrictions.
+ */
+template <typename T>
+struct fake_complex {
```
[the remark this answers is not in the document]
Comment (upsj) [reply cuda/base/types.hpp:69]: Hope that doesn't break CUDA 😄

File: common/unified/matrix/sparsity_csr_kernels.cpp (hunk 19 above)
[the remark this answers is not in the document]
Comment (thoasm) [reply common/unified/matrix/sparsity_csr_kernels.cpp:51]: Sounds good

File: reference/matrix/ell_kernels.cpp [hunk 21]
```diff
@@ -141,6 +141,47 @@ GKO_INSTANTIATE_FOR_EACH_MIXED_VALUE_AND_INDEX_TYPE(
...
+    for (size_type row = 0; row < output->get_size()[0]; row++) {
+        const auto row_begin = row_ptrs[row];
+        const auto row_end = row_ptrs[row + 1];
+        size_type col_idx = 0;
+        for (auto i = row_begin; i < row_end; i++) {
+            const auto entry = nonzeros.get_const_data()[i];
+            output->col_at(row, col_idx) = entry.column;
+            output->val_at(row, col_idx) = entry.value;
+            col_idx++;
+        }
```
[the remark this answers is not in the document]
Comment (upsj) [reply reference/matrix/ell_kernels.cpp:173]: I'd prefer to keep the loops simple/unsurprising, separate increments might get lost when reading the code.

File: reference/matrix/diagonal_kernels.cpp [hunk 22]
```diff
@@ -132,6 +132,24 @@ GKO_INSTANTIATE_FOR_EACH_VALUE_AND_INDEX_TYPE(
...
+template <typename ValueType, typename IndexType>
+void fill_in_matrix_data(
+    std::shared_ptr<const DefaultExecutor> exec,
+    const Array<matrix_data_entry<ValueType, IndexType>>& data,
+    matrix::Diagonal<ValueType>* output)
+{
+    for (size_type i = 0; i < data.get_num_elems(); i++) {
+        const auto entry = data.get_const_data()[i];
+        if (entry.row == entry.column) {
+            output->get_values()[entry.row] = entry.value;
```
[the remark this answers is not in the document]
Comment (upsj) [reply reference/matrix/diagonal_kernels.cpp:144]: It sounds like you're describing invalid matrix_data with out-of-bounds entries - should that be supported/caught? I'd probably defer that to device_matrix_data::validate() and #770

File: common/cuda_hip/matrix/fbcsr_kernels.hpp.inc (hunk 17 above)
Comment (upsj) [reply common/cuda_hip/matrix/fbcsr_kernels.hpp.inc:92]: Actually, I'll add tests and potential fixes in #904

File: core/matrix/diagonal.cpp (hunk 15 above)
[the remark this answers is not in the document]
Comment (yhmtsai) [reply core/matrix/diagonal.cpp:254]: I use the wrong function. it is the extract_diagonal

File: core/matrix/sparsity_csr.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (yhmtsai) [reply core/matrix/sparsity_csr.cpp:109]: Yes, I am thinking whether using the first value as the diagonal makes sense or not.
let's leave it as the original behavior now

File: include/ginkgo/core/matrix/fbcsr.hpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (yhmtsai) [reply include/ginkgo/core/matrix/fbcsr.hpp:308]: I see. it is a bug-fix, so we can change it

File: reference/matrix/diagonal_kernels.cpp (hunk 22 above)
[the remark this answers is not in the document]
Comment (yhmtsai) [reply reference/matrix/diagonal_kernels.cpp:144]: Maybe? it something like you did for core/test/utils/matrix_generator.hpp?

[the remark this answers is not in the document]
Comment (upsj) [reply reference/matrix/diagonal_kernels.cpp:144]: That was a bugfix for something that was broken earlier. To be honest, I think we should not have added a `read` function to Diagonal at all, since it is not a generic matrix format. But that was an earlier oversight, we can't change that until 2.0. Until then, I'd like to avoid having deal with device-side error handling

**Review by yhmtsai — APPROVED** [review]
LGTM. only some nit/comments

File: common/unified/matrix/ell_kernels.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (yhmtsai) [reply common/unified/matrix/ell_kernels.cpp]: is unpack_member the rename of fake_complex_unpack? Is there any difference between kernel_launch and type?

File: core/matrix/dense.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (yhmtsai) [reply core/matrix/dense.cpp:679]: what do you mean array view here? it does not use view here, I think?
In current develop, we do not adapt the input, either. Thus, it is fine now.
relative issue #96

[the remark this answers is not in the document]
Comment (upsj) [reply core/matrix/dense.cpp:679]: If the matrix already contains data in a view with a stride, this approach reuses the storage, which is probably closer to what users might expect?

File: common/cuda_hip/matrix/fbcsr_kernels.hpp.inc [hunk 23]
```diff
@@ -73,3 +73,86 @@ __global__ __launch_bounds__(default_block_size) void transpose_blocks(
...
+    Array<IndexType> row_idx_array{exec, num_blocks};
+    Array<size_type> block_ptr_array{exec, num_blocks};
+    col_idx_array.resize_and_reset(num_blocks);
+    value_array.resize_and_reset(num_blocks * bs * bs);
+    auto row_idxs = row_idx_array.get_data();
+    auto col_idxs = col_idx_array.get_data();
+    auto values = as_device_type(value_array.get_data());
+    auto block_ptrs = block_ptr_array.get_data();
+    auto block_ptr_it = thrust::device_pointer_cast(block_ptrs);
+    // write (block_row, block_col, block_begin) tuples for each block
```
[the remark this answers is not in the document]
Comment (upsj) [reply common/cuda_hip/matrix/fbcsr_kernels.hpp.inc:126]: it's the starting index of the block

File: omp/matrix/fbcsr_kernels.cpp [hunk 24]
```diff
@@ -150,6 +151,59 @@ GKO_INSTANTIATE_FOR_EACH_VALUE_AND_INDEX_TYPE(
...
+        }
+        if (new_block_col != block_col) {
+            // we encountered a new column, so insert it with block storage
+            col_idx_vec.emplace_back(new_block_col);
+            value_vec.resize(value_vec.size() + block_size * block_size);
+            block_col = new_block_col;
+        }
+        const auto local_row = entry.row % block_size;
+        const auto local_col = entry.column % block_size;
+        value_vec[row_ptrs_ptr[block_row] * block_size * block_size +
```
[the remark this answers is not in the document]
Comment (upsj) [reply omp/matrix/fbcsr_kernels.cpp:193]: Unfortunately, it wasn't entirely equivalent, I use the correct one in both places now

File: common/cuda_hip/matrix/fbcsr_kernels.hpp.inc [hunk 25]
```diff
@@ -73,3 +73,86 @@ __global__ __launch_bounds__(default_block_size) void transpose_blocks(
...
+                                         row_ptrs);
+    // fill in values
+    components::fill_array(exec, value_array.get_data(), num_blocks * bs * bs,
+                           zero<ValueType>());
+    thrust::for_each_n(
+        thrust::device, iota, num_blocks,
+        [block_ptrs, nnz, num_blocks, bs, data_ptr,
+         values] __device__(size_type i) {
+            const auto block_begin = block_ptrs[i];
+            const auto block_end = i < num_blocks ? block_ptrs[i + 1] : nnz;
```
[the remark this answers is not in the document]
Comment (upsj) [reply common/cuda_hip/matrix/fbcsr_kernels.hpp.inc:148]: You are right, it should be num_blocks - 1

File: common/unified/matrix/ell_kernels.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (upsj) [reply common/unified/matrix/ell_kernels.cpp]: One is device-specific (only for CUDA/HIP), the other one is generic in common kernels.

File: core/matrix/diagonal.cpp (hunk 15 above)
[the remark this answers is not in the document]
Comment (upsj) [reply core/matrix/diagonal.cpp:254]: I can't follow. A diagonal matrix is always square, so what would be the issue here?

File: dpcpp/components/device_matrix_data_kernels.dp.cpp [hunk 26]
```diff
@@ -0,0 +1,92 @@
...
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#include <oneapi/dpl/algorithm>
+#include <oneapi/dpl/execution>
```
[the remark this answers is not in the document]
Comment (upsj) [reply dpcpp/components/device_matrix_data_kernels.dp.cpp:36]: Good point, I forced its location now for the format_header script as well

File: common/cuda_hip/matrix/hybrid_kernels.hpp.inc [hunk 27]
```diff
@@ -140,3 +140,63 @@ __global__ __launch_bounds__(default_block_size) void add(
 
 
 }  // namespace kernel
+
+
+template <typename ValueType, typename IndexType>
+struct hybrid_tuple_unpack_functor {
```
[the remark this answers is not in the document]
Comment (upsj) [reply common/cuda_hip/matrix/hybrid_kernels.hpp.inc:146]: This is only necessary because a Lambda wouldn't be copy-assignable, which one of the algorithms requires. So this is only a local workaround without any generic usage.

File: core/matrix/hybrid.cpp [hunk 28]
```diff
@@ -195,60 +200,36 @@ void Hybrid<ValueType, IndexType>::move_to(Csr<ValueType, IndexType>* result)
...
-            }
-            ind++;
-        }
-        for (auto i = col; i < ell_lim; i++) {
-            tmp->ell_val_at(row, i) = zero<ValueType>();
-            tmp->ell_col_at(row, i) = 0;
-        }
+    auto exec = this->get_executor();
+    auto local_data = make_temporary_clone(exec, &data.nonzeros);
+    Array<int64> row_ptrs{exec, data.size[0] + 1};
```
[the remark this answers is not in the document]
Comment (upsj) [reply core/matrix/hybrid.cpp:207]: That is true, though I would separate between row pointers (indexing nonzeros, why I'm using int64 here) and row indices (indexing rows) here. The former are more likely to overflow than the latter.

File: core/matrix/diagonal.cpp (hunk 15 above)
[the remark this answers is not in the document]
Comment (yhmtsai) [reply core/matrix/diagonal.cpp:254]: extract_diagonal works for non-square matrix. do we also use this rule for reading?
